File: Python/src/bin/medium_49.py
# Daily Coding Problem: 49
# Difficulty: Medium
# Asked by: Amazon
# Authored by: TenthEdict

def max_subarray_sum(arr):
    if len(arr) == 0:
        return None
    
    current_sum = 0
    max_sum = 0

    for i in arr:
        current_sum = current_sum + i
        max_sum = max(max_sum, current_sum)
        if current_sum < 0:
            current_sum = 0

    return max_sum

def main():
    print(max_subarray_sum([34,-50,42,14,-5,86]))
    print(max_subarray_sum([-5,-1,-8,-9]))
    print(max_subarray_sum([1,2,3,4]))
    print(max_subarray_sum([]))
    print(max_subarray_sum([-2,1,-2,2,-1,2]))

main()

File: C/src/bin/medium_49.c
// Daily Coding Problem: 49
// Difficulty: Medium
// Asked by: Amazon
// Authored by: TenthEdict

#include <stdio.h>
#include <stdbool.h>

bool max_subarray_sum(int *arr, int len, int *result) {
    if (len <= 0) {
        *result = 0;
        return false;
    }

    int current_sum = 0;
    int max_sum = 0;

    for (int i = 0; i < len; i++)
    {
        current_sum += arr[i]; // issue 1
        max_sum = (max_sum > current_sum) ? max_sum : current_sum;
        current_sum = (current_sum < 0) ? 0 : current_sum;
    }

    *result = max_sum; // issue 3

    printf("%d\n", *result); // issue 2

    return true;
}

int main() {
    int test[] = {34, -50, 42, 14, -5, 86};
    int result;
    
    max_subarray_sum(test, sizeof(test) / sizeof(*test), &result);
    
    return 0;
}
